sql/src/main/java/io/crate/execution/engine/distribution/TransportDistributedResultAction.java

```diff
@@ -136,20 +142,23 @@ void pushResult(String node, DistributedResultRequest request, ActionListener<Di
     }
 
     private CompletableFuture<DistributedResultResponse> retryOrFailureResponse(DistributedResultRequest request,
-                                                                                int retry) {
+                                                                                @Nullable Iterator<TimeValue> retryDelay) {
 
-        if (retry > 20) {
-            return CompletableFutures.failedFuture(
-                new ContextMissingException(ContextMissingException.ContextType.JOB_EXECUTION_CONTEXT, request.jobId()));
-        } else {
-            int delay = (retry + 1) * 2;
+        if (retryDelay == null) {
+            retryDelay = RETRY_BACK_OFF.iterator();
+        }
+        if (retryDelay.hasNext()) {
+            TimeValue delay = retryDelay.next();
             if (logger.isTraceEnabled()) {
-                logger.trace("scheduling retry #{} to start node operation for jobId: {} in {}ms",
-                    retry, request.jobId(), delay);
+                logger.trace("scheduling retry to start node operation for jobId: {} in {}ms",
+                    request.jobId(), delay.getMillis());
             }
-            NodeOperationRunnable operationRunnable = new NodeOperationRunnable(request, retry);
-            scheduler.schedule(operationRunnable::run, delay, TimeUnit.MILLISECONDS);
+            NodeOperationRunnable operationRunnable = new NodeOperationRunnable(request, retryDelay);
+            scheduler.schedule(operationRunnable::run, delay.getMillis(), TimeUnit.MILLISECONDS);
             return operationRunnable;
+        } else {
+            return CompletableFutures.failedFuture(
+                new ContextMissingException(ContextMissingException.ContextType.JOB_EXECUTION_CONTEXT, request.jobId()));
```

Comment:
are we sure this is the only reason why this would continuously fail and break ? Can we maybe save the `NodeOperationRunnable` exception in case it completes exceptionally and use that?

Reply:
If the NodeOperationRunnable would fail for another reason, it wouldn't run into the retry logic but abort with the original exception.